### PR TITLE
spl: init at 0.3.2

### PR DESCRIPTION
**Pull request by TudbuT**
## Description of changes

Added SPL, a stack-based scripting language. Some SPL code can be found at https://git.tudbut.de/TudbuT/mkfunction/src/branch/main/mkpolynomial.spl . Repository is at https://git.tudbut.de/tudbut/spl . 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/sp/spl/package.nix [hunk 1]
```diff
@@ -0,0 +1,29 @@
+{
+  pkgs,
```
Comment (donovanglover) [inline pkgs/by-name/sp/spl/package.nix:2]:
```suggestion
  lib,
  fetchgit,
  buildRustPackage,
```

File: pkgs/by-name/sp/spl/package.nix [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+}:
+let
+  version = "0.3.1";
+  src = pkgs.fetchgit {
+    url = "https://git.tudbut.de/tudbut/spl";
+    rev = "v${version}";
+    hash = "sha256-Bj3voUwyqUWg1f86l/Vob+kb5jJUZmf0ZG9IxyYjgh0=";
+  };
+in
```
Comment (donovanglover) [inline pkgs/by-name/sp/spl/package.nix:11]:
```suggestion
```

Unnecessary `let`

File: pkgs/by-name/sp/spl/package.nix [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+pkgs.rustPlatform.buildRustPackage {
```
Comment (donovanglover) [inline pkgs/by-name/sp/spl/package.nix:12]:
```suggestion
rustPlatform.buildRustPackage rec {
```

File: pkgs/by-name/sp/spl/package.nix [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  pname = "spl";
+  version = version;
+
+  src = src;
+
+  cargoLock = {
+    lockFile = "${src}/Cargo.lock";
+  };
```
Comment (donovanglover) [inline pkgs/by-name/sp/spl/package.nix:20]:
```suggestion
```
This is an [import from derivation](https://nix.dev/manual/nix/2.24/language/import-from-derivation), which is not allowed in nixpkgs. `cargoHash` should be used if possible.

File: pkgs/by-name/sp/spl/package.nix [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+  meta = {
+    description = "Stack Programming Language";
```
Comment (donovanglover) [inline pkgs/by-name/sp/spl/package.nix:23]:
```suggestion
    description = "Simple, concise, concatenative scripting language";
```
The description should describe the program. Read the [guidelines](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#conventions) for more information.

**Review by donovanglover — COMMENTED** [review]
Welcome to nixpkgs! Make sure to follow the [Rust conventions](https://nixos.org/manual/nixpkgs/stable/#compiling-rust-applications-with-cargo) in the nixpkgs manual to get started.

File: pkgs/by-name/sp/spl/package.nix (hunk 3 above)
Comment (TudbuT) [reply pkgs/by-name/sp/spl/package.nix:12]: Interesting! I did originally use rec, but a friend had told me it was slow and not recommended for nixpkgs. I assume they were wrong then?

Anyways, I'll fix it shortly.

**Review by Aleksanaa — APPROVED** [review]

**Comment by Aleksanaa** [comment]
Waiting for darwin build

File: pkgs/by-name/sp/spl/package.nix (hunk 3 above)
Comment (donovanglover) [reply pkgs/by-name/sp/spl/package.nix]: I'm unaware of how significant the performance impact is but there are significantly more Rust packages using `rec` than `let ... in`.

Generally I think following the Rust documentation helps make the code base more consistent for everyone, and a treewide change would be ideal if there was a consensus on changing this (as well as a documentation update).

**Comment by TudbuT** [comment]
Is it normal for that darwin build to take this long?

**Comment by Aleksanaa** [comment]
Oh sorry, at least x86_64-darwin seems okay, less unlikely to cause a build failure

**Pull request merged**